From 216a8b765794bdfe96ca2518b3eed482cf34e797 Mon Sep 17 00:00:00 2001
From: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
Date: Tue, 21 Jul 2026 10:53:14 +0530
Subject: [PATCH 5/6] pwm: core: add support for legacy PWM drivers

- Add compatibility support for drivers implementing legacy PWM ops.
- Introduce PWM chip allocation and chip-private data helpers.
- Validate PWM states and improve handling of invalid transitions.

Signed-off-by: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
---
 drivers/pwm/core.c  | 242 +++++++++++++++++++++++++++++++++++++++++---
 include/linux/pwm.h |  39 +++++++
 2 files changed, 265 insertions(+), 16 deletions(-)

diff --git a/drivers/pwm/core.c b/drivers/pwm/core.c
index 108d8c98b16f..453668d650b5 100644
--- a/drivers/pwm/core.c
+++ b/drivers/pwm/core.c
@@ -139,6 +139,25 @@ static void pwm_apply_debug(struct pwm_device *pwm,
 	}
 }
 
+static bool pwm_state_valid(const struct pwm_state *state)
+{
+        /*
+         * For a disabled state all other state description is irrelevant and
+         * and supposed to be ignored. So also ignore any strange values and
+         * consider the state ok.
+         */
+        if (!state->enabled)
+                return true;
+
+        if (!state->period)
+                return false;
+
+        if (state->duty_cycle > state->period)
+                return false;
+
+        return true;
+}
+
 /**
  * __pwm_apply() - atomically apply a new state to a PWM device
  * @pwm: PWM device
@@ -149,31 +168,98 @@ static int __pwm_apply(struct pwm_device *pwm, const struct pwm_state *state)
 	struct pwm_chip *chip;
 	int err;
 
-	if (!pwm || !state || !state->period ||
-	    state->duty_cycle > state->period)
+	if (!pwm || !state)
 		return -EINVAL;
 
+	if (!pwm_state_valid(state)) {
+		/*
+		 * Allow to transition from one invalid state to another.
+		 * This ensures that you can e.g. change the polarity while
+		 * the period is zero. (This happens on stm32 when the hardware
+		 * is in its poweron default state.) This greatly simplifies
+		 * working with the sysfs API where you can only change one
+		 * parameter at a time.
+		 */
+		if (!pwm_state_valid(&pwm->state)) {
+			pwm->state = *state;
+			return 0;
+		}
+
+		return -EINVAL;
+	}
+
 	chip = pwm->chip;
 
 	if (state->period == pwm->state.period &&
-	    state->duty_cycle == pwm->state.duty_cycle &&
-	    state->polarity == pwm->state.polarity &&
-	    state->enabled == pwm->state.enabled &&
-	    state->usage_power == pwm->state.usage_power)
+			state->duty_cycle == pwm->state.duty_cycle &&
+			state->polarity == pwm->state.polarity &&
+			state->enabled == pwm->state.enabled &&
+			state->usage_power == pwm->state.usage_power)
 		return 0;
 
-	err = chip->ops->apply(chip, pwm, state);
-	trace_pwm_apply(pwm, state, err);
-	if (err)
-		return err;
+	if (chip->ops->apply) {
+		err = chip->ops->apply(chip, pwm, state);
+		trace_pwm_apply(pwm, state, err);
+		if (err)
+			return err;
 
-	pwm->state = *state;
+		pwm->state = *state;
 
-	/*
-	 * only do this after pwm->state was applied as some
-	 * implementations of .get_state depend on this
-	 */
-	pwm_apply_debug(pwm, state);
+		/*
+		 * only do this after pwm->state was applied as some
+		 * implementations of .get_state depend on this
+		 */
+		pwm_apply_debug(pwm, state);
+	} else {
+		/*
+		 * FIXME: restore the initial state in case of error.
+		 */
+		if (state->polarity != pwm->state.polarity) {
+			if (!chip->ops->set_polarity)
+				return -EINVAL;
+
+			/*
+			 * Changing the polarity of a running PWM is
+			 * only allowed when the PWM driver implements
+			 * ->apply().
+			 */
+			if (pwm->state.enabled) {
+				chip->ops->disable(chip, pwm);
+				pwm->state.enabled = false;
+			}
+
+			err = chip->ops->set_polarity(chip, pwm,
+					state->polarity);
+			if (err)
+				return err;
+
+			pwm->state.polarity = state->polarity;
+		}
+
+		if (state->period != pwm->state.period ||
+				state->duty_cycle != pwm->state.duty_cycle) {
+			err = chip->ops->config(pwm->chip, pwm,
+					state->duty_cycle,
+					state->period);
+			if (err)
+				return err;
+
+			pwm->state.duty_cycle = state->duty_cycle;
+			pwm->state.period = state->period;
+		}
+
+		if (state->enabled != pwm->state.enabled) {
+			if (state->enabled) {
+				err = chip->ops->enable(chip, pwm);
+				if (err)
+					return err;
+			} else {
+				chip->ops->disable(chip, pwm);
+			}
+
+			pwm->state.enabled = state->enabled;
+		}
+	}
 
 	return 0;
 }
@@ -475,10 +561,134 @@ static void of_pwmchip_remove(struct pwm_chip *chip)
 		of_node_put(pwmchip_parent(chip)->of_node);
 }
 
+static inline struct pwm_chip *pwmchip_from_dev(struct device *pwmchip_dev)
+{
+        return dev_get_drvdata(pwmchip_dev);
+}
+
+#define PWMCHIP_ALIGN ARCH_DMA_MINALIGN
+
+static void *pwmchip_priv(struct pwm_chip *chip)
+{
+	return (void *)chip + ALIGN(sizeof(*chip) + chip->npwm * sizeof(*chip->pwms), PWMCHIP_ALIGN);
+}
+
+/* This is the counterpart to pwmchip_alloc() */
+void pwmchip_put(struct pwm_chip *chip)
+{
+        put_device(chip->dev);
+}
+EXPORT_SYMBOL_GPL(pwmchip_put);
+
+static void pwmchip_release(struct device *pwmchip_dev)
+{
+        struct pwm_chip *chip = pwmchip_from_dev(pwmchip_dev);
+
+        kfree(chip);
+}
+
+struct pwm_chip *pwmchip_alloc(struct device *parent, unsigned int npwm, size_t sizeof_priv)
+{
+	struct pwm_chip *chip;
+	struct device *pwmchip_dev;
+	size_t alloc_size;
+	unsigned int i;
+
+	alloc_size = size_add(sizeof(*chip) + npwm * sizeof(*chip->pwms),
+			sizeof_priv);
+
+	chip = kzalloc(alloc_size, GFP_KERNEL);
+	if (!chip)
+		return ERR_PTR(-ENOMEM);
+
+	chip->npwm = npwm;
+	chip->uses_pwmchip_alloc = true;
+
+	pwmchip_dev = chip->dev;
+	device_initialize(pwmchip_dev);
+	pwmchip_dev->parent = parent;
+	pwmchip_dev->release = pwmchip_release;
+
+	pwmchip_set_drvdata(chip, pwmchip_priv(chip));
+
+	for (i = 0; i < chip->npwm; i++) {
+		struct pwm_device *pwm = &chip->pwms[i];
+		pwm->chip = chip;
+		pwm->hwpwm = i;
+	}
+
+	return chip;
+}
+EXPORT_SYMBOL_GPL(pwmchip_alloc);
+
+static void devm_pwmchip_put(void *data)
+{
+        struct pwm_chip *chip = data;
+
+        pwmchip_put(chip);
+}
+
+struct pwm_chip *devm_pwmchip_alloc(struct device *parent, unsigned int npwm, size_t sizeof_priv)
+{
+	struct pwm_chip *chip;
+	int ret;
+
+	chip = pwmchip_alloc(parent, npwm, sizeof_priv);
+	if (IS_ERR(chip))
+		return chip;
+
+	ret = devm_add_action_or_reset(parent, devm_pwmchip_put, chip);
+	if (ret)
+		return ERR_PTR(ret);
+
+	return chip;
+}
+EXPORT_SYMBOL_GPL(devm_pwmchip_alloc);
+
+
+/**
+ * pwm_set_chip_data() - set private chip data for a PWM
+ * @pwm: PWM device
+ * @data: pointer to chip-specific data
+ *
+ * Returns: 0 on success or a negative error code on failure.
+ */
+int pwm_set_chip_data(struct pwm_device *pwm, void *data)
+{
+	if (!pwm)
+		return -EINVAL;
+
+	pwm->chip_data = data;
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(pwm_set_chip_data);
+
+/**
+ * pwm_get_chip_data() - get private chip data for a PWM
+ * @pwm: PWM device
+ *
+ * Returns: A pointer to the chip-private data for the PWM device.
+ */
+void *pwm_get_chip_data(struct pwm_device *pwm)
+{
+	return pwm ? pwm->chip_data : NULL;
+}
+EXPORT_SYMBOL_GPL(pwm_get_chip_data);
+
 static bool pwm_ops_check(const struct pwm_chip *chip)
 {
 	const struct pwm_ops *ops = chip->ops;
 
+	/* driver supports legacy, non-atomic operation */
+	if (ops->config && ops->enable && ops->disable) {
+		if (IS_ENABLED(CONFIG_PWM_DEBUG))
+			dev_warn(pwmchip_parent(chip),
+					"Driver needs updating to atomic API\n");
+
+		return true;
+	}
+
 	if (!ops->apply)
 		return false;
 
diff --git a/include/linux/pwm.h b/include/linux/pwm.h
index 07af6910bdce..39aa9d4a8962 100644
--- a/include/linux/pwm.h
+++ b/include/linux/pwm.h
@@ -2,6 +2,7 @@
 #ifndef __LINUX_PWM_H
 #define __LINUX_PWM_H
 
+#include <linux/device.h>
 #include <linux/err.h>
 #include <linux/mutex.h>
 #include <linux/of.h>
@@ -70,6 +71,7 @@ struct pwm_state {
  * @flags: flags associated with the PWM device
  * @hwpwm: per-chip relative index of the PWM device
  * @chip: PWM chip providing this PWM device
+ * @chip_data: chip-private data associated with the PWM device
  * @args: PWM arguments
  * @state: last applied state
  * @last: last implemented state (for PWM_DEBUG)
@@ -79,6 +81,7 @@ struct pwm_device {
 	unsigned long flags;
 	unsigned int hwpwm;
 	struct pwm_chip *chip;
+	void *chip_data;
 
 	struct pwm_args args;
 	struct pwm_state state;
@@ -251,6 +254,11 @@ struct pwm_capture {
  * @get_state: get the current PWM state. This function is only
  *	       called once per PWM device when the PWM chip is
  *	       registered.
+ * @owner: helps prevent removal of modules exporting active PWMs
+ * @config: configure duty cycles and period length for this PWM
+ * @set_polarity: configure the polarity of this PWM
+ * @enable: enable PWM output toggling
+ * @disable: disable PWM output toggling
  */
 struct pwm_ops {
 	int (*request)(struct pwm_chip *chip, struct pwm_device *pwm);
@@ -261,31 +269,47 @@ struct pwm_ops {
 		     const struct pwm_state *state);
 	int (*get_state)(struct pwm_chip *chip, struct pwm_device *pwm,
 			 struct pwm_state *state);
+	struct module *owner;
+
+	/* Only used by legacy drivers */
+	int (*config)(struct pwm_chip *chip, struct pwm_device *pwm,
+			int duty_ns, int period_ns);
+	int (*set_polarity)(struct pwm_chip *chip, struct pwm_device *pwm,
+			enum pwm_polarity polarity);
+	int (*enable)(struct pwm_chip *chip, struct pwm_device *pwm);
+	void (*disable)(struct pwm_chip *chip, struct pwm_device *pwm);
+
 };
 
 /**
  * struct pwm_chip - abstract a PWM controller
  * @dev: device providing the PWMs
  * @ops: callbacks for this PWM controller
+ * @base: number of first PWM controlled by this chip
  * @owner: module providing this chip
  * @id: unique number of this PWM chip
  * @npwm: number of PWMs controlled by this chip
+ * @of_pwm_n_cells: number of cells expected in the device tree PWM specifier
  * @of_xlate: request a PWM device given a device tree PWM specifier
  * @atomic: can the driver's ->apply() be called in atomic context
+ * @uses_pwmchip_alloc: signals if pwmchip_allow was used to allocate this chip
  * @pwms: array of PWM devices allocated by the framework
  */
 struct pwm_chip {
 	struct device *dev;
 	const struct pwm_ops *ops;
+	int base;
 	struct module *owner;
 	unsigned int id;
 	unsigned int npwm;
 
 	struct pwm_device * (*of_xlate)(struct pwm_chip *chip,
 					const struct of_phandle_args *args);
+	unsigned int of_pwm_n_cells;
 	bool atomic;
 
 	/* only used internally by the PWM framework */
+	bool uses_pwmchip_alloc;
 	struct pwm_device *pwms;
 };
 
@@ -294,6 +318,16 @@ static inline struct device *pwmchip_parent(const struct pwm_chip *chip)
 	return chip->dev;
 }
 
+static inline void *pwmchip_get_drvdata(struct pwm_chip *chip)
+{
+        return dev_get_drvdata(chip->dev);
+}
+
+static inline void pwmchip_set_drvdata(struct pwm_chip *chip, void *data)
+{
+        dev_set_drvdata(chip->dev, data);
+}
+
 #if IS_ENABLED(CONFIG_PWM)
 /* PWM user APIs */
 int pwm_apply_might_sleep(struct pwm_device *pwm, const struct pwm_state *state);
@@ -382,7 +416,12 @@ static inline bool pwm_might_sleep(struct pwm_device *pwm)
 /* PWM provider APIs */
 int pwm_capture(struct pwm_device *pwm, struct pwm_capture *result,
 		unsigned long timeout);
+void pwmchip_put(struct pwm_chip *chip);
+struct pwm_chip *pwmchip_alloc(struct device *parent, unsigned int npwm, size_t sizeof_priv);
+struct pwm_chip *devm_pwmchip_alloc(struct device *parent, unsigned int npwm, size_t sizeof_priv);
 
+int pwm_set_chip_data(struct pwm_device *pwm, void *data);
+void *pwm_get_chip_data(struct pwm_device *pwm);
 int __pwmchip_add(struct pwm_chip *chip, struct module *owner);
 #define pwmchip_add(chip) __pwmchip_add(chip, THIS_MODULE)
 void pwmchip_remove(struct pwm_chip *chip);
-- 
2.25.1

